2-multitenant/modules/env_baseline/db.tf

```
@@ -0,0 +1,70 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# CloudSQL Postgres instance
+module "cloudsql" {
+  source  = "GoogleCloudPlatform/sql-db/google//modules/postgresql"
+  version = "~> 20.0"
+
+  for_each = data.google_compute_subnetwork.default
+
+  project_id = local.cluster_project_id
+  region     = each.value.region
+
+  name                = "db-${each.value.region}-${var.env}"
+  database_version    = "POSTGRES_14"
+  enable_default_db   = false
+  tier                = "db-custom-1-3840"
+  deletion_protection = false
+  availability_type   = "REGIONAL"
+
+  additional_databases = [
+    {
+      name      = "accounts-db"
+      charset   = ""
+      collation = ""
+    },
+    {
+      name      = "ledger-db"
+      charset   = ""
+      collation = ""
+    }
+  ]
+  user_name     = "admin"
+  user_password = "admin" # this is a security risk - do not do this for real world use-cases!
+}
+
+resource "google_service_account" "bank_of_anthos" {
+  project      = local.cluster_project_id
+  account_id   = "bank-of-anthos"
+  display_name = "bank-of-anthos"
+}
+
+resource "google_project_iam_member" "bank_of_anthos" {
+  for_each = toset(["roles/cloudsql.client", "roles/cloudsql.instanceUser"])
+  project  = local.cluster_project_id
+  role     = each.value
+  member   = "serviceAccount:${google_service_account.bank_of_anthos.email}"
+}
+
+resource "google_service_account_iam_binding" "workload_identity" {
+  service_account_id = google_service_account.bank_of_anthos.name
+  role               = "roles/iam.workloadIdentityUser"
```

Should this be the cluster project?

```suggestion
  project  = local.cluster_project_id
  role               = "roles/iam.workloadIdentityUser"
```

---

it does not take a 'project' as input, instead, google_service_account.bank_of_anthos.name is the full resource name with 'project'.

Error: Error applying IAM policy for service account 'projects/eab-gke-development-1609/serviceAccounts/bank-of-anthos@eab-gke-development-1609.iam.gserviceaccount.com': Error setting IAM policy for service account 'projects/eab-gke-development-1609/serviceAccounts/bank-of-anthos@eab-gke-development-1609.iam.gserviceaccount.com': googleapi: Error 400: Identity Pool does not exist (eab-gke-development-1609.svc.id.goog). Please check that you specified a valid resource name as returned in the `name` attribute in the configuration API., badRequest

probably this a timing issue ? it takes time to create identity pool ?

---

Hey @yliaog - Good point, didn't double check that particular resource, unfortunately that it doesn't include `project`  The error I was trying to troubleshoot was the `sqladmin.googleapis.com` API wasn't enabled on the `ci-enterprise-application` project, and rather than adding the API, I was hoping to force the use of a specific project rather than relying on the provider default project.

For the current error, my initial recommendation would be adding an explicit dependency that workload identity has been enabled in the project.  As that's currently done by the GKE module, one option might be to depend on (or use as it should default to `[project_id].svc.id.goog`) the value of `module.gke[*].identity_namespace`.

---

module.gke[each.key].identity_namespace or module.gke[*].identity_namespace both caused the errors below.

│   on db.tf line 71, in resource "google_service_account_iam_binding" "workload_identity":
│   71:     module.gke[each.key].identity_namespace
│ 
│ A single static variable reference is required: only attribute access and
│ indexing with constant keys. No calculations, function calls, template
│ expressions, etc are allowed here.

---

Correct - You can only use `each.key` if you are using a `for_each` in the resource.

I believe you have two options if using `depends_on`, either depend on all clusters (don't need to include the `[*]`) with `depends_on = [ module.gke.identity_namespace ]`, or just a single cluster (which is all that as needed as workload identity is enabled project wide) with `depends_on = [ module.gke[0].identity_namespace ]`.

I suspect you could also avoid the `depends_on`, and instead use `${module.gke[0].identity_namespace}` to replace `${local.cluster_project_id}.svc.id.goog`. (Need to double check the returned value.)